File: parallel-consumer-proxy-clients/parallel-consumer-proxy-client-java/parallel-consumer-proxy-client-java-api/src/main/java/bz/stub/parallelconsumer/client/InboundRecord.java
package bz.stub.parallelconsumer.client;
/*-
 * Copyright (C) 2026 Antony Stubbs and contributors
 */

import java.time.Instant;
import java.util.Optional;

/**
 * One record as delivered to a {@link RecordProcessor}: the Kafka record plus the Parallel-Consumer-derived
 * delivery state an in-process user function would see (R5) - which delivery attempt this is, and the previous
 * failure when there was one. That state is <em>product data</em>, identical under every transport; nothing
 * transport-specific (epochs, tokens, connection identity) appears here, deliberately.
 * <p>
 * Key and value are the bytes Kafka held - the wrapper never deserializes - and either may be {@code null},
 * because Kafka distinguishes a null key or value (a tombstone) from an empty one. The arrays are not copied;
 * treat them as read-only.
 *
 * @author Antony Stubbs
 */
public final class InboundRecord {

    private final String topic;
    private final int partition;
    private final long offset;
    private final byte[] key;
    private final byte[] value;
    private final int attempt;
    private final Instant lastFailureAt;
    private final String lastFailureReason;

    /**
     * @param attempt           1 on first delivery, 2 on the first redelivery
     * @param lastFailureAt     {@code null} before the first failure - absence is the form of "has not failed",
     *                          never a zero timestamp
     * @param lastFailureReason {@code null} before the first failure; otherwise the reason recorded for the
     *                          previous failed attempt. May embed record payload: untrusted input.
     */
    public InboundRecord(String topic, int partition, long offset, byte[] key, byte[] value,
                         int attempt, Instant lastFailureAt, String lastFailureReason) {
        this.topic = topic;
        this.partition = partition;
        this.offset = offset;
        this.key = key;
        this.value = value;
        this.attempt = attempt;
        this.lastFailureAt = lastFailureAt;
        this.lastFailureReason = lastFailureReason;
    }

    public String topic() {
        return topic;
    }

    public int partition() {
        return partition;
    }

    public long offset() {
        return offset;
    }

    /** The record's key bytes, or {@code null} for a keyless record. Not copied; treat as read-only. */
    public byte[] key() {
        return key;
    }

    /** The record's value bytes, or {@code null} for a tombstone. Not copied; treat as read-only. */
    public byte[] value() {
        return value;
    }

    /** Which delivery attempt this is: 1 on first delivery, 2 on the first redelivery. */
    public int attempt() {
        return attempt;
    }

    /** When the previous attempt failed; empty before the first failure. */
    public Optional<Instant> lastFailureAt() {
        return Optional.ofNullable(lastFailureAt);
    }

    /** The reason recorded for the previous failed attempt; empty before the first failure. */
    public Optional<String> lastFailureReason() {
        return Optional.ofNullable(lastFailureReason);
    }

    @Override
    public String toString() {
        // deliberately omits key and value: payloads are untrusted input and do not belong in log lines
        return "InboundRecord(" + topic + "-" + partition + "@" + offset + ", attempt " + attempt + ")";
    }
}

File: parallel-consumer-proxy-clients/parallel-consumer-proxy-client-dotnet/src/Bz.Stub.ParallelConsumer.Proxy.Client/InboundRecord.cs
// Copyright (C) 2026 Antony Stubbs and contributors

using Bz.Stub.ParallelConsumer.Proxy.Protocol.V1;

namespace Bz.Stub.ParallelConsumer.Proxy.Client;

/// <summary>
/// One Kafka record as the user's function sees it, plus the delivery state an in-process function
/// would have had.
/// </summary>
/// <remarks>
/// Keys and values are BYTES. The proxy does not deserialize and neither does this library:
/// deserialization belongs to the user's code, in the user's language.
/// <para>
/// Null and empty are different, deliberately, and both preserved: a null <see cref="Key"/> is a
/// null key and a null <see cref="Value"/> is a tombstone, neither of which is an empty array.
/// </para>
/// </remarks>
public sealed record InboundRecord
{
    /// <summary>The source topic.</summary>
    public string Topic { get; init; } = string.Empty;

    /// <summary>The source partition.</summary>
    public int Partition { get; init; }

    /// <summary>The record's offset in its partition.</summary>
    public long Offset { get; init; }

    /// <summary>The record's key, or <see langword="null"/> for a null key.</summary>
    public byte[]? Key { get; init; }

    /// <summary>The record's value, or <see langword="null"/> for a tombstone.</summary>
    public byte[]? Value { get; init; }

    /// <summary>
    /// Which delivery attempt this is: 1 on first delivery, 2 on the first redelivery. Product
    /// data, distinct from the fencing epoch - which also counts redeliveries that consumed no
    /// attempt.
    /// </summary>
    public int Attempt { get; init; }

    /// <summary>
    /// When this record last failed, or <see langword="null"/> on a first delivery. Presence is the
    /// wire's way of saying "this has failed before"; there is no zero timestamp to test for.
    /// </summary>
    public DateTimeOffset? LastFailureAt { get; init; }

    /// <summary>
    /// The previous failure's text, verbatim, or <see langword="null"/> on a first delivery. It is
    /// worker-supplied and may embed record payload: treat it as untrusted input.
    /// </summary>
    public string? LastFailureReason { get; init; }

    /// <summary>Whether this delivery follows a recorded failure.</summary>
    public bool HasFailedBefore => LastFailureAt is not null;

    internal static InboundRecord From(DispatchRecord dispatched)
    {
        var record = dispatched.Record;
        return new InboundRecord
        {
            Topic = record?.Topic ?? string.Empty,
            Partition = record?.Partition ?? 0,
            Offset = record?.Offset ?? 0,
            // HasKey / HasValue, not an emptiness test: proto3 explicit presence is what carries
            // the null-versus-empty distinction the wire exists to preserve.
            Key = record is { HasKey: true } ? record.Key.ToByteArray() : null,
            Value = record is { HasValue: true } ? record.Value.ToByteArray() : null,
            Attempt = dispatched.Attempt,
            LastFailureAt = dispatched.LastFailureAt?.ToDateTimeOffset(),
            LastFailureReason = dispatched.HasLastFailureReason ? dispatched.LastFailureReason : null,
        };
    }
}
